tests/test_simpleGenericMap.py

```diff
@@ -42,58 +42,84 @@ def makeMap(mapType, values):
 
     def testInitKeywords(self):
-        for (keyType, target) in self.targets:
-            self.checkInitKwargs(target, self.getTestData(keyType), msg=str(target))
+        for target in self.targets:
+            for keyType in self.getValidKeys(target):
+                self.checkInitKwargs(target, self.getTestData(keyType), msg=str(target))
 
     def testInitPairs(self):
-        for (keyType, target) in self.targets:
-            self.checkInitPairs(keyType, target, self.getTestData(keyType), msg=str(target))
+        for target in self.targets | self.implicitKeyTargets:
+            for keyType in self.getValidKeys(target):
+                self.checkInitPairs(target, self.getTestData(keyType), msg=str(target))
 
     def testInitMapping(self):
-        for (keyType, target) in self.targets:
-            # Init from dict
-            self.checkInitMapping(keyType, target, self.getTestData(keyType), msg=str(target))
-            # Init from GenericMap
-            self.checkInitMapping(keyType, target, self.makeMap(target, self.getTestData(keyType)),
-                                  msg=str(target))
+        for target in self.targets | self.implicitKeyTargets:
+            for keyType in self.getValidKeys(target):
+                # Init from dict
+                self.checkInitMapping(target, self.getTestData(keyType), msg=str(target))
+                # Init from GenericMap
+                self.checkInitMapping(target, target(self.getTestData(keyType)),
+                                      msg=str(target))
+
+    def testUnknownKeys(self):
+        for target in self.implicitKeyTargets:
+            with self.assertRaises(TypeError):
+                target()
+            # Should not raise
+            target(dtype=str)
+
+    def testMixedKeys(self):
+        badData = {"What is the product of six and eight?": "Ultimate Question",
+                   42: "Ultimate Answer",
+                   }
+        for target in self.targets | self.implicitKeyTargets:
+            with self.assertRaises(TypeError):
+                target(badData)
+            with self.assertRaises(TypeError):
+                target(badData.items())
+        for target in self.targets:
+            with self.assertRaises(TypeError):
+                target(**badData)
 
     def testFromKeys(self):
-        for (keyType, target) in self.targets:
-            keys = self.getTestData(keyType).keys()
-            for value in self.getTestData(keyType).values():
-                self.checkFromKeys(keyType, target, keys, value,
-                                   msg=" class=%s, value=%r" % (target, value))
-            self.checkFromKeysDefault(keyType, target, keys, msg=" class=%s, no value" % (target))
+        for target in self.targets:
+            for keyType in self.getValidKeys(target):
+                keys = self.getTestData(keyType).keys()
+                for value in self.getTestData(keyType).values():
```

Comment:
Triply nested loops in tests make me nervous. I don't know if there's a better approach, but this tends to make debugging failed tests more difficult.

Reply:
I provide the `msg` argument to the assertions for exactly that reason. I think it's fine as-is.